[SPARK-38047][K8S] Add `OUTLIER_NO_FALLBACK` executor roll policy

alexholmes opened the pull request:
### What changes were proposed in this pull request?

This PR aims to add a new executor roll policy which allows users to skip rolling in cases where there are no outlier executors.

### Why are the changes needed?

As currently implemented an executor is always rolled every `spark.kubernetes.executor.rollInterval` interval. In environments where starting of executors can introduce latencies it may be desirable for users to have the option to determine if rolling should only happen when outliers are found.

### Does this PR introduce _any_ user-facing change?

No, this is an additional option being added to a new feature in Apache Spark 3.3.

### How was this patch tested?

Pass the CIs with the newly added test cases.

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun reviewed (commented):
Welcome to the Apache Spark community, @alexholmes .

dongjoon-hyun commented:
Let me close and reopen this to trigger your GitHub Action properly.

alexholmes commented:
Thank you @dongjoon-hyun !

dongjoon-hyun commented:
It seems that your GitHub Action is not responding still for some reason.
- https://github.com/alexholmes/spark/actions/workflows/build_and_test.yml

dongjoon-hyun commented:
Could you do force-push this on your branch?

dongjoon-hyun commented:
FYI, I revised your PR title according to the community style, `[SPARK-38047][K8S]`. Each Apache community has different style unfortunately.

dongjoon-hyun reviewed (changes requested):
BTW, I'd like to recommend not to touch the other policy, @alexholmes .
They are orthogonal to the outlier. Please make a new policy for your requirement.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -182,6 +182,16 @@ private[spark] object Config extends Logging {
+      .doc("Only roll an executor if it is an outlier. An outlier is defined to be at least two " +
+        "standard deviation outside of the mean. If no outlier is found then no executor will " +
+        "be rolled."
+      )
```
Please place `)` at line 189.

alexholmes replied:
I removed the config so I'm going to resolve this.

alexholmes committed:
SPARK-38047 - Provide an option to only roll executors if they are outliers

alexholmes committed:
Makes a couple of changes:

1. Introduced a new policy, OUTLIER_OR_TOTAL_DURATION, which provides a fallback of TOTAL_DURATION policy for outliers (basically the prior implementation of OUTLIER).
2. Updated the OUTLIER policy to not provide a fallback, therefore allowing no executors to be rolled when there are no outliners

alexholmes commented:
@dongjoon-hyun I've updated the PR with the following changes:

1. Introduced a new policy, OUTLIER_OR_TOTAL_DURATION, which provides a fallback of TOTAL_DURATION policy for outliers (basically the prior implementation of OUTLIER).
2. Updated the OUTLIER policy to not provide a fallback, therefore allowing no executors to be rolled when there are no outliners

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -165,7 +166,9 @@ private[spark] object Config extends Logging {
         "total task time, total task GC time, and the number of failed tasks if exists. " +
-        "If there is no outlier, it works like TOTAL_DURATION policy.")
+        "If there is no outlier then no executor will be rolled." +
+        "OUTLIER_OR_TOTAL_DURATION policy picks an outlier using the OUTLIER policy above. " +
```
Could you keep the previous behavior by making `OUTLIER_OR_TOTAL_DURATION` as the default?
```scala
- .createWithDefault(ExecutorRollPolicy.OUTLIER.toString)
+ .createWithDefault(ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION.toString) 
```

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
-          outliers(listWithoutDriver, e => e.totalGCTime) ++
-          outliers(listWithoutDriver, e => e.failedTasks) ++
-          listWithoutDriver.sortBy(_.totalDuration).reverse
+      case ExecutorRollPolicy.OUTLIER => outliers(listWithoutDriver)
```
If you don't mind, please put `outliers(listWithoutDriver)` at the next line.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
+  /**
+   * We build multiple outlier lists and concat in the following importance order to find
+   * outliers in various perspective:
+   * AVERAGE_DURATION > TOTAL_DURATION > TOTAL_GC_TIME > FAILED_TASKS
```
Could you add two more leading space here?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
+   * AVERAGE_DURATION > TOTAL_DURATION > TOTAL_GC_TIME > FAILED_TASKS
+   * Since we will choose only first item, the duplication is okay.
+   */
+  private def outliers(listWithoutDriver: Seq[v1.ExecutorSummary]) =
```
`outliers` -> `outliersFromMultipleDimentions`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
    * In case of normal distribution, this is known to be 2.5 percent roughly.
    */
-  private def outliers(
+  private def outliersForFn(
```
Let's keep the original method name because this is focusing on a single dimension outlier.

dongjoon-hyun reviewed (commented):
Could you revise the PR title and description according to the new PR content?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -147,7 +147,8 @@ private[spark] object Config extends Logging {
   object ExecutorRollPolicy extends Enumeration {
-    val ID, ADD_TIME, TOTAL_GC_TIME, TOTAL_DURATION, AVERAGE_DURATION, FAILED_TASKS, OUTLIER = Value
+    val ID, ADD_TIME, TOTAL_GC_TIME, TOTAL_DURATION, AVERAGE_DURATION, FAILED_TASKS,
+    OUTLIER, OUTLIER_OR_TOTAL_DURATION = Value
```
indentation?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -165,7 +166,9 @@ private[spark] object Config extends Logging {
         "at least two standard deviation from the mean in average task time, " +
         "total task time, total task GC time, and the number of failed tasks if exists. " +
-        "If there is no outlier, it works like TOTAL_DURATION policy.")
+        "If there is no outlier then no executor will be rolled." +
```
We need a space at the end of the string. `rolled. "`

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPlugin.scala:
```diff
@@ -117,29 +117,36 @@ class ExecutorRollDriverPlugin extends DriverPlugin with Logging {
+  private def outliers(listWithoutDriver: Seq[v1.ExecutorSummary]) =
+    outliersForFn(
+      listWithoutDriver.filter(_.totalTasks > 0),
+      e => e.totalDuration / e.totalTasks) ++
```
Please merge line 135 ~ 137 as a one liner if possible.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:
```diff
@@ -132,7 +132,9 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
-    ExecutorRollPolicy.values.foreach { value =>
+    ExecutorRollPolicy.values
+      .filter(_ != ExecutorRollPolicy.OUTLIER)
+      .foreach { value =>
```
The following would be shorter and reasonable.
```scala
ExecutorRollPolicy.values.filter(_ != ExecutorRollPolicy.OUTLIER).foreach { value =>
```

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:
```diff
@@ -172,10 +174,53 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
+      plugin.invokePrivate(_choose(list, ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION)))
+  }
+
+  test("Policy: OUTLIER_OR_TOTAL_DURATION - Detect an average task duration outlier") {
```
Actually, this is orthogonal to your proposal. We usually put one theme in a PR and avoid mixing different things.
Could you make a separate PR for adding more test coverage for `OUTLIER_OR_TOTAL_DURATION`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorRollPluginSuite.scala:
```diff
@@ -172,10 +174,53 @@ class ExecutorRollPluginSuite extends SparkFunSuite with PrivateMethodTester {
+      plugin.invokePrivate(_choose(list :+ outlier, ExecutorRollPolicy.OUTLIER_OR_TOTAL_DURATION)))
+  }
+
+  test("Policy: OUTLIER - Return None if not outliers") {
```
`not outliers` -> `there is no outlier`.

dongjoon-hyun replied:
Please move this test case after `test("A one-item executor list")`.

alexholmes committed:
Additional changes:

1. Kept the behavior of OUTLIERS policy unchanged, along with its associated unit tests.
2. Introduced a new OUTLIER_NO_FALLBACK policy which doesn't fallback to TOTAL_DURATION in cases where there are no outliers

alexholmes committed:
code formatting to address feedback

alexholmes commented:
@dongjoon-hyun Thanks for your feedback. I've made the following changes:

1. Kept the behavior of OUTLIERS policy unchanged, along with its associated unit tests.
2. Introduced a new OUTLIER_NO_FALLBACK policy which doesn't fallback to TOTAL_DURATION in cases where there are no outliers
3. Updated the title and content of this PR to reflect these changes
4. Addressed your formatting comments above

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you for updating, @alexholmes .
Merged to master for Apache Spark 3.3.0.
